src/parser/sparqlParser/SparqlQleverVisitor.cpp

```diff
@@ -927,6 +927,57 @@ GraphPatternOperation Visitor::visitSpatialQuery(
   return spatialQuery;
 }
 
+GraphPatternOperation Visitor::visitWordSearchQuery(
+    Parser::ServiceGraphPatternContext* ctx) {
+  auto parseWordSearchQuery =
+      [ctx](parsedQuery::WordSearchQuery& textSearchQuery,
+            const parsedQuery::GraphPatternOperation& op) {
+        if (std::holds_alternative<parsedQuery::BasicGraphPattern>(op)) {
+          textSearchQuery.addBasicPattern(
+              std::get<parsedQuery::BasicGraphPattern>(op));
+        } else {
+          reportError(
+              ctx,
+              "Unsupported element in wordSearchQuery."
+              "wordSearchQuery may only consist of triples for configuration");
+        }
+      };
+
+  parsedQuery::GraphPattern graphPattern = visit(ctx->groupGraphPattern());
+  parsedQuery::WordSearchQuery wordSearchQuery;
+  for (const auto& op : graphPattern._graphPatterns) {
+    parseWordSearchQuery(wordSearchQuery, op);
+  }
+
+  return wordSearchQuery;
+}
+
+GraphPatternOperation Visitor::visitEntitySearchQuery(
+    Parser::ServiceGraphPatternContext* ctx) {
+  auto parseEntitySearchQuery = [ctx](parsedQuery::EntitySearchQuery&
+                                          textSearchQuery,
+                                      const parsedQuery::GraphPatternOperation&
+                                          op) {
```

Comment:
Can you confirm, that basically all our magic service visitors (2 preexisting plus those you add) use basically exactly the same code, so we can reduce the code duplication by introducing a common template (or even a helper function that uses a reference to the base class, as the `MagicServiceQuery` uses inheritance and virtual functions.